### DISCUSSION TITLE
Pulling test run data using API for README.md shields.io badges

### DISCUSSION THREAD

**Discussion by yyahav** [ludeknovy/jtl-reporter#267] [Ideas]
Would be nice if I could pull data from the `/scenarios/<scenario_name>` endpoint using the API, such as:
- Last test run P90 value
- Last test run errors count
- Last test run <... whatever custom data the user configures ...>

To be used with `img.shields.io/endpoint` to display a badge in README.md.

### Example
<img width="645" alt="image" src="https://github.com/ludeknovy/jtl-reporter/assets/88027106/02915a7d-c9a3-4a07-97a0-d06986cbe457">

### Required JSON Endpoint Response (according to `shields.io`)
```
{ "schemaVersion": 1, "label": "<scenario_name>", "message": "<customizable KPI>", "color": "<customizable color>" }
```

I wouldn't mind contributing and open PR for that, I just need someone to point me in the right direction to the API and project admin settings.

Here's a badge example with the JTLReporter logo:
[![JTLReporter Badge](https://img.shields.io/endpoint?style=for-the-badge&color=blue&url=https://gist.githubusercontent.com/yyahav/6ec1c08e0813635623b11dfc9d1edda0/raw)](https://github.com/ludeknovy/jtl-reporter)

**Comment by ludeknovy** [comment]
Hi @yyahav 
So the data you ask for is actually already there, and you can fetch it using API, you just need to generate a new api token: https://jtlreporter.site/docs/guides/administration/api-token

To have it return the response in the required format, there would be changes needed of course. My first thought would be to reuse the scenario trends endpoint https://github.com/ludeknovy/jtl-reporter-be/blob/master/src/server/controllers/scenario/trends/get-scenario-trends-controller.ts#L11

Actually, does the schields.io support auth headers? Without it, it's not going to work.

**Reply by yyahav** [reply]
You're right... it won't work with the auth requirement...

Maybe it's possible to create a specific endpoint for that purpose alone, that gets the scenario id in the GET request and doesn't require the auth headers?
Wdyt?

**Reply by ludeknovy** [reply]
Sorry, my bad. The above mentioned endpoint also accepts a token as param, that skips the auth header validation - see https://github.com/ludeknovy/jtl-reporter/discussions/240

So the request made: GET /scenarios/My%20Scenario/trends?environment=&token=<token>

So a similar approach could be used to return only one property for the last test run. Wdyt?

**Reply by yyahav** [reply]
All good =)
Yes, it sounds right and won't require much effort. Its controller is in the same area as the scenario trends endpoint, right?
So the only extra info I need is the location of the UI so I can add the customization options.

Thank you so much!

**Reply by ludeknovy** [reply]
Yes, the new controller should be in the same folder as the scenario trends endpoint.
Regarding the UI, a lazy option is not to do any changes, as the token generated could be reused for the new endpoint. The more advanced option would be to adjust the form with purpose selector- but that would also require more work on BE- to store the token purpose to differentiate them in the scenario token listing. Honestly, I think the first option is good enough if proper documentation will be provided.

**Reply by yyahav** [reply]
Just to see if I understand correctly - no changes to UI means that the endpoint will respond with hard coded specific KPIs (label is the scenario name and color can be overridden in the shields.io call), use the same token as the Scenario Share token, and no user customization will be available.

I guess that as a starter of such a feature that's ok, and later on could be developed further for customization.

**Reply by ludeknovy** [reply]
Yes, no changes to UI.
You can make the customization based on the parameters, eg: metric=p90 or metric=errorCount, etc.. that will return you the desired data.

**Reply by yyahav** [reply]
Hopefully I'll have some time to add this endpoint soon.
I'll create a PR after testing it.

Thank you!

**Reply by ludeknovy** [reply]
Ok. Let me know if you need any help with this.
